src/core/consumerthread.h

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CONSUMERTHREAD_H
+#define BITCOIN_CONSUMERTHREAD_H
+
+#include <thread>
+#include <core/producerconsumerqueue.h>
+
+template<WorkerMode MODE>
+class ConsumerThread;
+
+//! A WorkItem() encapsulates a task that can be processed by a ConsumerThread()
+//! @see ConsumerThread()
+template<WorkerMode MODE>
+class WorkItem
+{
+    friend ConsumerThread<MODE>; //<! only a consumer thread can execute a WorkItem
+
+protected:
+    WorkItem(){};
+    virtual void operator()(){};
+};
+
+//! A special WorkItem() that is used to interrupt a blocked ConsumerThread() so that it can terminate
+template<WorkerMode MODE>
+class ShutdownPill : public WorkItem<MODE> {
+    friend ConsumerThread<MODE>;
+
+private:
+    ShutdownPill(ConsumerThread<MODE>& consumer) : m_consumer(consumer) {};
```

Comment:
Yeah - it is pretty complicated. I didn't want to poke through the queue api just to enable shutdown - you'd have to have `T Pop()` potentially not return a `T` (i.e. throw an exception) - which seemed less desirable and maybe equally complicated.

Having said that - a lot of the complexity here is introduced to handle:
- being able to shutdown a specific ConsumerThread without shutting down others (in reality you probably only ever want to shut down all of them when you terminate the process)
- allowing for a queue with fewer slots than the number of threads servicing it (unlikely)
Only reason I allowed for these is so that the APIs to the queue work the way they sound like they should work and for unit test completeness. If I discard the above the code gets much simpler and it will just be broken for two use cases that seem pretty unlikely to ever happen right now (but you never know - and then maybe somebody gets frustrated one day).